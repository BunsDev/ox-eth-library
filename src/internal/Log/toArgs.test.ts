import { Abi, AbiEvent, Hex, Log } from 'ox'
import { expect, test } from 'vitest'
import { anvilMainnet } from '../../../test/anvil.js'
import { address } from '../../../test/constants/addresses.js'

test('default', () => {
  const abiEvent = AbiEvent.from(
    'event Transfer(address indexed from, address indexed to, uint256 value)',
  )
  const args = Log.toArgs(
    {
      data: '0x0000000000000000000000000000000000000000000000000000000000000001',
      topics: [
        '0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef',
        '0x000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac',
        '0x000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac',
      ],
    },
    {
      abiEvent,
    },
  )
  expect(args).toMatchInlineSnapshot(`
    {
      "from": "0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac",
      "to": "0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac",
      "value": 1n,
    }
  `)
})

test('behavior: filter logs, network', async () => {
  const abi = Abi.from([
    'event Transfer(address indexed from, address indexed to, uint256 value)',
    'event Approval(address indexed owner, address indexed spender, uint256 value)',
  ])
  const logs = await anvilMainnet.request({
    method: 'eth_getLogs',
    params: [
      {
        fromBlock: Hex.fromNumber(
          (anvilMainnet.config.forkBlockNumber as bigint) - 5n,
        ),
        toBlock: Hex.fromNumber(anvilMainnet.config.forkBlockNumber!),
      },
    ],
  })
  expect(
    logs
      .map((rpcLog) => {
        try {
          const log = Log.fromRpc(rpcLog)
          const abiEvent = AbiEvent.fromAbi(abi, { name: log.topics[0] })
          return {
            args: Log.toArgs(log, {
              abiEvent,
              matchArgs: {
                from: [
                  '0x916fa3c896bb0bb9171a3f5bfb1f76b5cff3b46b',
                  '0x2aeee741fa1e21120a21e57db9ee545428e683c9',
                ],
              },
            }),
            event: abiEvent.name,
            log,
          }
        } catch {
          return null
        }
      })
      .filter(Boolean),
  ).toMatchSnapshot()
})

test('behavior: args, named', () => {
  const abiEvent = AbiEvent.from(
    'event Transfer(address indexed from, address indexed to, uint256 value)',
  )

  expect(
    Log.toArgs(
      {
        data: '0x0000000000000000000000000000000000000000000000000000000000000001',
        topics: [
          '0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef',
          '0x000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac',
          '0x000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac',
        ],
      },
      {
        abiEvent,
        matchArgs: {
          from: '0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac',
          to: '0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac',
          value: 1n,
        },
      },
    ),
  ).toMatchInlineSnapshot(`
    {
      "from": "0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac",
      "to": "0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac",
      "value": 1n,
    }
  `)
  expect(
    Log.toArgs(
      {
        data: '0x0000000000000000000000000000000000000000000000000000000000000001',
        topics: [
          '0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef',
          '0x000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac',
          '0x000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac',
        ],
      },
      {
        abiEvent,
        matchArgs: {
          from: '0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac',
        },
      },
    ),
  ).toMatchInlineSnapshot(`
    {
      "from": "0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac",
      "to": "0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac",
      "value": 1n,
    }
  `)
  expect(
    Log.toArgs(
      {
        data: '0x0000000000000000000000000000000000000000000000000000000000000001',
        topics: [
          '0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef',
          '0x000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac',
          '0x000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac',
        ],
      },
      {
        abiEvent,
        matchArgs: {
          from: ['0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac', address.vitalik],
          to: null,
        },
      },
    ),
  ).toMatchInlineSnapshot(`
    {
      "from": "0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac",
      "to": "0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac",
      "value": 1n,
    }
  `)
})

test('behavior: args, unnamed', () => {
  const abiEvent = AbiEvent.from(
    'event Transfer(address indexed, address indexed, uint256)',
  )

  expect(
    Log.toArgs(
      {
        data: '0x0000000000000000000000000000000000000000000000000000000000000001',
        topics: [
          '0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef',
          '0x000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac',
          '0x000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac',
        ],
      },
      {
        abiEvent,
        matchArgs: [
          '0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac',
          '0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac',
          1n,
        ],
      },
    ),
  ).toMatchInlineSnapshot(`
    [
      "0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac",
      "0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac",
      1n,
    ]
  `)
  expect(
    Log.toArgs(
      {
        data: '0x0000000000000000000000000000000000000000000000000000000000000001',
        topics: [
          '0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef',
          '0x000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac',
          '0x000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac',
        ],
      },
      {
        abiEvent,
        matchArgs: ['0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac'],
      },
    ),
  ).toMatchInlineSnapshot(`
    [
      "0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac",
      "0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac",
      1n,
    ]
  `)
  expect(
    Log.toArgs(
      {
        data: '0x0000000000000000000000000000000000000000000000000000000000000001',
        topics: [
          '0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef',
          '0x000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac',
          '0x000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac',
        ],
      },
      {
        abiEvent,
        matchArgs: [
          null,
          ['0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac', address.vitalik],
          1n,
        ],
      },
    ),
  ).toMatchInlineSnapshot(`
    [
      "0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac",
      "0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac",
      1n,
    ]
  `)
})

test('error: args mismatch, named', () => {
  const abiEvent = AbiEvent.from(
    'event Transfer(address indexed from, address indexed to, uint256 value)',
  )

  expect(() =>
    Log.toArgs(
      {
        data: '0x0000000000000000000000000000000000000000000000000000000000000001',
        topics: [
          '0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef',
          '0x000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac',
          '0x000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac',
        ],
      },
      {
        abiEvent,
        matchArgs: {
          from: '0xa5cc3c03994db5b0d9a5eedd10cabab0813678ad',
          to: '0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac',
          value: 1n,
        },
      },
    ),
  ).toThrowErrorMatchingInlineSnapshot(`
    [Log.ArgsMismatchError: Provided arguments to not match the arguments decoded from the log.

    Event: event Transfer(address indexed from, address indexed to, uint256 value)
    Decoded Arguments: 
      from:   0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac
      to:     0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac
      value:  1
    Provided Arguments: 
      from:   0xa5cc3c03994db5b0d9a5eedd10cabab0813678ad
      to:     0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac
      value:  1]
  `)
})

test('error: args mismatch, unnamed', () => {
  const abiEvent = AbiEvent.from(
    'event Transfer(address indexed, address indexed, uint256)',
  )

  expect(() =>
    Log.toArgs(
      {
        data: '0x0000000000000000000000000000000000000000000000000000000000000001',
        topics: [
          '0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef',
          '0x000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac',
          '0x000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac',
        ],
      },
      {
        abiEvent,
        matchArgs: ['0xa5cc3c03994db5b0d9a5eedd10cabab0813678ad'],
      },
    ),
  ).toThrowErrorMatchingInlineSnapshot(`
    [Log.ArgsMismatchError: Provided arguments to not match the arguments decoded from the log.

    Event: event Transfer(address indexed, address indexed, uint256)
    Decoded Arguments: 
      0:  0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac
      1:  0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac
      2:  1
    Provided Arguments: 
      0:  0xa5cc3c03994db5b0d9a5eedd10cabab0813678ad]
  `)
})

test('error: args mismatch, string', () => {
  const abiEvent = AbiEvent.from('event Transfer(string indexed a)')

  expect(() =>
    Log.toArgs(
      {
        topics: [
          '0x7cebee4ee226a36ff8751d9d69bb8265f5138c825f8c25d7ebdd60d972ffe5be',
          '0x1c8aff950685c2ed4bc3174f3472287b56d9517b9c948127319a09a7a36deac8',
        ],
      },
      {
        abiEvent,
        matchArgs: {
          a: '0xdeadbeef',
        },
      },
    ),
  ).toThrowErrorMatchingInlineSnapshot(`
    [Log.ArgsMismatchError: Provided arguments to not match the arguments decoded from the log.

    Event: event Transfer(string indexed a)
    Decoded Arguments: 
      a:  0x1c8aff950685c2ed4bc3174f3472287b56d9517b9c948127319a09a7a36deac8
    Provided Arguments: 
      a:  0xdeadbeef]
  `)
})

test('error: input not found, named', () => {
  const abiEvent = AbiEvent.from(
    'event Transfer(address indexed from, address indexed to, uint256 value)',
  )

  expect(() =>
    Log.toArgs(
      {
        data: '0x0000000000000000000000000000000000000000000000000000000000000001',
        topics: [
          '0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef',
          '0x000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac',
          '0x000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac',
        ],
      },
      {
        abiEvent,
        matchArgs: {
          // @ts-expect-error
          a: 'b',
          from: '0xa5cc3c03994db5b0d9a5eedd10cabab0813678ad',
          to: '0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac',
          value: 1n,
        },
      },
    ),
  ).toThrowErrorMatchingInlineSnapshot(
    `[AbiEvent.InputNotFoundError: Parameter "a" not found on \`event Transfer(address indexed from, address indexed to, uint256 value)\`.]`,
  )
})

test('error: input not found, unnamed', () => {
  const abiEvent = AbiEvent.from(
    'event Transfer(address indexed, address indexed, uint256)',
  )

  expect(() =>
    Log.toArgs(
      {
        data: '0x0000000000000000000000000000000000000000000000000000000000000001',
        topics: [
          '0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef',
          '0x000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac',
          '0x000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac',
        ],
      },
      {
        abiEvent,
        matchArgs: [
          '0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac',
          '0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac',
          1n,
          // @ts-expect-error
          1n,
        ],
      },
    ),
  ).toThrowErrorMatchingInlineSnapshot(
    `[AbiEvent.InputNotFoundError: Parameter "3" not found on \`event Transfer(address indexed, address indexed, uint256)\`.]`,
  )
})

test('error: provided args, but no actual event args', () => {
  const abiEvent = AbiEvent.from('event Transfer()')

  expect(() =>
    Log.toArgs(
      {
        topics: [
          '0x406dade31f7ae4b5dbc276258c28dde5ae6d5c2773c5745802c493a2360e55e0',
        ],
      },
      {
        abiEvent,
        matchArgs: {
          // @ts-ignore
          from: '0xa5cc3c03994db5b0d9a5eedd10cabab0813678az',
          to: '0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac',
          value: 1n,
        },
      },
    ),
  ).toThrowErrorMatchingInlineSnapshot(`
    [Log.ArgsMismatchError: Provided arguments to not match the arguments decoded from the log.

    Event: event Transfer()
    Decoded Arguments: None
    Provided Arguments: 
      from:   0xa5cc3c03994db5b0d9a5eedd10cabab0813678az
      to:     0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac
      value:  1]
  `)
})
