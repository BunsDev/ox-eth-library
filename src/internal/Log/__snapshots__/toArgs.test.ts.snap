// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`behavior: filter logs, network 1`] = `
[
  {
    "args": {
      "from": "0x2aeee741fa1e21120a21e57db9ee545428e683c9",
      "to": "0xf3de3c0d654fda23dad170f0f320a92172509127",
      "value": 108534933194540417n,
    },
    "event": "Transfer",
    "log": {
      "address": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
      "blockHash": "0xc350d807505fb835650f0013632c5515592987ba169bbc6626d9fc54d91f0f0b",
      "blockNumber": 19868015n,
      "data": "0x000000000000000000000000000000000000000000000000018197f25ac80581",
      "logIndex": 3,
      "removed": false,
      "topics": [
        "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
        "0x0000000000000000000000002aeee741fa1e21120a21e57db9ee545428e683c9",
        "0x000000000000000000000000f3de3c0d654fda23dad170f0f320a92172509127",
      ],
      "transactionHash": "0xcdd096880f66c302c214338b8f860f39757aa10bc5f14561b21a42be88ef3f6a",
      "transactionIndex": 0,
    },
  },
  {
    "args": {
      "from": "0x2aeee741fa1e21120a21e57db9ee545428e683c9",
      "to": "0x6b75d8af000000e20b7a7ddf000ba900b4009a80",
      "value": 97392377756385280n,
    },
    "event": "Transfer",
    "log": {
      "address": "0xc56c7a0eaa804f854b536a5f3d5f49d2ec4b12b8",
      "blockHash": "0xc350d807505fb835650f0013632c5515592987ba169bbc6626d9fc54d91f0f0b",
      "blockNumber": 19868015n,
      "data": "0x000000000000000000000000000000000000000000000000015a01da00000000",
      "logIndex": 9,
      "removed": false,
      "topics": [
        "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
        "0x0000000000000000000000002aeee741fa1e21120a21e57db9ee545428e683c9",
        "0x0000000000000000000000006b75d8af000000e20b7a7ddf000ba900b4009a80",
      ],
      "transactionHash": "0xb2fc82757d06e42ea83b56371c87836b5445b7d810fde0d3a28680df13a8922e",
      "transactionIndex": 1,
    },
  },
  {
    "args": {
      "from": "0x2aeee741fa1e21120a21e57db9ee545428e683c9",
      "to": "0x3fc91a3afd70395cd496c647d5a6cc9d4b2b7fad",
      "value": 112911865374245030n,
    },
    "event": "Transfer",
    "log": {
      "address": "0xc56c7a0eaa804f854b536a5f3d5f49d2ec4b12b8",
      "blockHash": "0xc350d807505fb835650f0013632c5515592987ba169bbc6626d9fc54d91f0f0b",
      "blockNumber": 19868015n,
      "data": "0x000000000000000000000000000000000000000000000000019124be4a8f00a6",
      "logIndex": 19,
      "removed": false,
      "topics": [
        "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
        "0x0000000000000000000000002aeee741fa1e21120a21e57db9ee545428e683c9",
        "0x0000000000000000000000003fc91a3afd70395cd496c647d5a6cc9d4b2b7fad",
      ],
      "transactionHash": "0x353fdfc38a2f26115daadee9f5b8392ce62b84f410957967e2ed56b35338cdd0",
      "transactionIndex": 2,
    },
  },
  {
    "args": {
      "from": "0x2aeee741fa1e21120a21e57db9ee545428e683c9",
      "to": "0x6b75d8af000000e20b7a7ddf000ba900b4009a80",
      "value": 116567388098396160n,
    },
    "event": "Transfer",
    "log": {
      "address": "0xc56c7a0eaa804f854b536a5f3d5f49d2ec4b12b8",
      "blockHash": "0xc350d807505fb835650f0013632c5515592987ba169bbc6626d9fc54d91f0f0b",
      "blockNumber": 19868015n,
      "data": "0x000000000000000000000000000000000000000000000000019e216c00000000",
      "logIndex": 31,
      "removed": false,
      "topics": [
        "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
        "0x0000000000000000000000002aeee741fa1e21120a21e57db9ee545428e683c9",
        "0x0000000000000000000000006b75d8af000000e20b7a7ddf000ba900b4009a80",
      ],
      "transactionHash": "0x9d4150a317a7509cd71051c90479e79452a1c32eb2c773345d02031c5aecf03b",
      "transactionIndex": 4,
    },
  },
  {
    "args": {
      "from": "0x2aeee741fa1e21120a21e57db9ee545428e683c9",
      "to": "0x3fc91a3afd70395cd496c647d5a6cc9d4b2b7fad",
      "value": 109148138375907580n,
    },
    "event": "Transfer",
    "log": {
      "address": "0xc56c7a0eaa804f854b536a5f3d5f49d2ec4b12b8",
      "blockHash": "0xc350d807505fb835650f0013632c5515592987ba169bbc6626d9fc54d91f0f0b",
      "blockNumber": 19868015n,
      "data": "0x0000000000000000000000000000000000000000000000000183c5a74fca88fc",
      "logIndex": 47,
      "removed": false,
      "topics": [
        "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
        "0x0000000000000000000000002aeee741fa1e21120a21e57db9ee545428e683c9",
        "0x0000000000000000000000003fc91a3afd70395cd496c647d5a6cc9d4b2b7fad",
      ],
      "transactionHash": "0xe92e98715972ac47714058dd73d0c1d8ede4c414d1e8e4f784d2f40a1b071fc5",
      "transactionIndex": 5,
    },
  },
  {
    "args": {
      "from": "0x2aeee741fa1e21120a21e57db9ee545428e683c9",
      "to": "0x6b75d8af000000e20b7a7ddf000ba900b4009a80",
      "value": 1736695203235889152n,
    },
    "event": "Transfer",
    "log": {
      "address": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
      "blockHash": "0xc350d807505fb835650f0013632c5515592987ba169bbc6626d9fc54d91f0f0b",
      "blockNumber": 19868015n,
      "data": "0x0000000000000000000000000000000000000000000000001819fb1600000000",
      "logIndex": 60,
      "removed": false,
      "topics": [
        "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
        "0x0000000000000000000000002aeee741fa1e21120a21e57db9ee545428e683c9",
        "0x0000000000000000000000006b75d8af000000e20b7a7ddf000ba900b4009a80",
      ],
      "transactionHash": "0x8f94f5e869f972f21602bcce30f6117bfa6c14ab23bf83f0bf84ef071362c3cc",
      "transactionIndex": 6,
    },
  },
  {
    "args": {
      "from": "0x2aeee741fa1e21120a21e57db9ee545428e683c9",
      "to": "0x5c9321e92ba4eb43f2901c4952358e132163a85a",
      "value": 203109041221092900n,
    },
    "event": "Transfer",
    "log": {
      "address": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
      "blockHash": "0xc350d807505fb835650f0013632c5515592987ba169bbc6626d9fc54d91f0f0b",
      "blockNumber": 19868015n,
      "data": "0x00000000000000000000000000000000000000000000000002d19698cca74624",
      "logIndex": 76,
      "removed": false,
      "topics": [
        "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
        "0x0000000000000000000000002aeee741fa1e21120a21e57db9ee545428e683c9",
        "0x0000000000000000000000005c9321e92ba4eb43f2901c4952358e132163a85a",
      ],
      "transactionHash": "0x55062567d775d5668c0b68b7773cd175d8153c05428e4eb36e39b006718ddc34",
      "transactionIndex": 10,
    },
  },
  {
    "args": {
      "from": "0x916fa3c896bb0bb9171a3f5bfb1f76b5cff3b46b",
      "to": "0x070f380daf4b33fe5015e1f97acc60ec95225692",
      "value": 2000000000000000000000n,
    },
    "event": "Transfer",
    "log": {
      "address": "0xa9e8acf069c58aec8825542845fd754e41a9489a",
      "blockHash": "0xc350d807505fb835650f0013632c5515592987ba169bbc6626d9fc54d91f0f0b",
      "blockNumber": 19868015n,
      "data": "0x00000000000000000000000000000000000000000000006c6b935b8bbd400000",
      "logIndex": 346,
      "removed": false,
      "topics": [
        "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
        "0x000000000000000000000000916fa3c896bb0bb9171a3f5bfb1f76b5cff3b46b",
        "0x000000000000000000000000070f380daf4b33fe5015e1f97acc60ec95225692",
      ],
      "transactionHash": "0xcb2c1e2a1eebf7e68065a7c4c87c4479184d958b2a917dedc9566d7488e634bb",
      "transactionIndex": 83,
    },
  },
  {
    "args": {
      "from": "0x2aeee741fa1e21120a21e57db9ee545428e683c9",
      "to": "0x3fc91a3afd70395cd496c647d5a6cc9d4b2b7fad",
      "value": 80957567181972519n,
    },
    "event": "Transfer",
    "log": {
      "address": "0xc56c7a0eaa804f854b536a5f3d5f49d2ec4b12b8",
      "blockHash": "0x7ff6ab6fa5cdc5d926ddc5317731a17177bac7d7fda7d0ac923673924e4a84c6",
      "blockNumber": 19868016n,
      "data": "0x000000000000000000000000000000000000000000000000011f9e7a79c2e427",
      "logIndex": 15,
      "removed": false,
      "topics": [
        "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
        "0x0000000000000000000000002aeee741fa1e21120a21e57db9ee545428e683c9",
        "0x0000000000000000000000003fc91a3afd70395cd496c647d5a6cc9d4b2b7fad",
      ],
      "transactionHash": "0xc23af402abae2512fd00aa2d1d6ee7e336d19a0147f9d7c4798fa8a051abff5d",
      "transactionIndex": 21,
    },
  },
  {
    "args": {
      "from": "0x2aeee741fa1e21120a21e57db9ee545428e683c9",
      "to": "0x536ea5a25a8b6a7ca426634748861fb1b9b00fff",
      "value": 373361223546994169n,
    },
    "event": "Transfer",
    "log": {
      "address": "0xc56c7a0eaa804f854b536a5f3d5f49d2ec4b12b8",
      "blockHash": "0xfe0dbe4e500fa68a39d0fccc4650834856f705515d80ce733fab1985c2431cf6",
      "blockNumber": 19868017n,
      "data": "0x000000000000000000000000000000000000000000000000052e720dfe6c7df9",
      "logIndex": 53,
      "removed": false,
      "topics": [
        "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
        "0x0000000000000000000000002aeee741fa1e21120a21e57db9ee545428e683c9",
        "0x000000000000000000000000536ea5a25a8b6a7ca426634748861fb1b9b00fff",
      ],
      "transactionHash": "0x50ea38a331c35a10a5bc64a806f5c332978d151ddb7d56dd42f2f2fa88742820",
      "transactionIndex": 8,
    },
  },
  {
    "args": {
      "from": "0x2aeee741fa1e21120a21e57db9ee545428e683c9",
      "to": "0xa8e1aed437f83898784fdc5c0256412f10be33ae",
      "value": 753208683940175675n,
    },
    "event": "Transfer",
    "log": {
      "address": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
      "blockHash": "0xfe0dbe4e500fa68a39d0fccc4650834856f705515d80ce733fab1985c2431cf6",
      "blockNumber": 19868017n,
      "data": "0x0000000000000000000000000000000000000000000000000a73ef4eaffe5f3b",
      "logIndex": 172,
      "removed": false,
      "topics": [
        "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
        "0x0000000000000000000000002aeee741fa1e21120a21e57db9ee545428e683c9",
        "0x000000000000000000000000a8e1aed437f83898784fdc5c0256412f10be33ae",
      ],
      "transactionHash": "0x9a87eab93587ff6847cc66c110e24ede8d1a6be4f8f55fe27ad1db7660d2d06d",
      "transactionIndex": 25,
    },
  },
  {
    "args": {
      "from": "0x2aeee741fa1e21120a21e57db9ee545428e683c9",
      "to": "0x3fc91a3afd70395cd496c647d5a6cc9d4b2b7fad",
      "value": 456137500000000001n,
    },
    "event": "Transfer",
    "log": {
      "address": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
      "blockHash": "0xfe0dbe4e500fa68a39d0fccc4650834856f705515d80ce733fab1985c2431cf6",
      "blockNumber": 19868017n,
      "data": "0x000000000000000000000000000000000000000000000000065486a3b83c1801",
      "logIndex": 614,
      "removed": false,
      "topics": [
        "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
        "0x0000000000000000000000002aeee741fa1e21120a21e57db9ee545428e683c9",
        "0x0000000000000000000000003fc91a3afd70395cd496c647d5a6cc9d4b2b7fad",
      ],
      "transactionHash": "0x4310d1d8edc016a87fc14febb11cb2dc4a95a6446d1271c0578e9cd3cc0a37a9",
      "transactionIndex": 288,
    },
  },
  {
    "args": {
      "from": "0x2aeee741fa1e21120a21e57db9ee545428e683c9",
      "to": "0xe76014c179f19da26bb30a0f085ff0a466b92829",
      "value": 170476114935531789n,
    },
    "event": "Transfer",
    "log": {
      "address": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
      "blockHash": "0x28d779fee521840d1b80b3a457259e34910e8ec871ca75f90e1f7d0a1a8e846b",
      "blockNumber": 19868018n,
      "data": "0x000000000000000000000000000000000000000000000000025da71f916bc90d",
      "logIndex": 9,
      "removed": false,
      "topics": [
        "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
        "0x0000000000000000000000002aeee741fa1e21120a21e57db9ee545428e683c9",
        "0x000000000000000000000000e76014c179f19da26bb30a0f085ff0a466b92829",
      ],
      "transactionHash": "0x04985f7e5636d0178f0e18b62ccfb5a07b64138822a124908e80e70039014893",
      "transactionIndex": 2,
    },
  },
  {
    "args": {
      "from": "0x2aeee741fa1e21120a21e57db9ee545428e683c9",
      "to": "0x3fc91a3afd70395cd496c647d5a6cc9d4b2b7fad",
      "value": 42817549093540788n,
    },
    "event": "Transfer",
    "log": {
      "address": "0xc56c7a0eaa804f854b536a5f3d5f49d2ec4b12b8",
      "blockHash": "0xa93b995575bda48d4cf45a4f72593a48a744786b5e32e5ff92a21372f7a60875",
      "blockNumber": 19868019n,
      "data": "0x00000000000000000000000000000000000000000000000000981e5583df93b4",
      "logIndex": 141,
      "removed": false,
      "topics": [
        "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
        "0x0000000000000000000000002aeee741fa1e21120a21e57db9ee545428e683c9",
        "0x0000000000000000000000003fc91a3afd70395cd496c647d5a6cc9d4b2b7fad",
      ],
      "transactionHash": "0xf17d415c1393ab67869b4233c8b957abd36fd22f2e99d4fff501ec4172e33e34",
      "transactionIndex": 33,
    },
  },
  {
    "args": {
      "from": "0x2aeee741fa1e21120a21e57db9ee545428e683c9",
      "to": "0x2396588b50d4d9a6c75295423c1d732de5d7c574",
      "value": 77149588706135041n,
    },
    "event": "Transfer",
    "log": {
      "address": "0xc56c7a0eaa804f854b536a5f3d5f49d2ec4b12b8",
      "blockHash": "0x0c3e51fc62e1dd7401c8884882755b72e1c5720d0b01c75443cdfa8a129d3fc9",
      "blockNumber": 19868020n,
      "data": "0x0000000000000000000000000000000000000000000000000112172465658c01",
      "logIndex": 56,
      "removed": false,
      "topics": [
        "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
        "0x0000000000000000000000002aeee741fa1e21120a21e57db9ee545428e683c9",
        "0x0000000000000000000000002396588b50d4d9a6c75295423c1d732de5d7c574",
      ],
      "transactionHash": "0xd14bf1505356f83e55d4baffa315e0e9210479dff7486c4164a050c935d5bccf",
      "transactionIndex": 11,
    },
  },
]
`;
